kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+     *  - enctypes=...      -- key-salt types
+     *  - revoke=true       -- delete old keys (concrete princs only)
+     *  - rotate=true       -- change keys (no-op for virtual princs)
+     *  - create=true       -- create a concrete princ
+     *  - ro=true           -- perform no writes
+     */
+
+    if (strcmp(key, "realm") == 0 && val) {
+        if (!r->realm && !(r->realm = strdup(val)))
+            ret = krb5_enomem(r->context);
+    } else if (strcmp(key, "materialize") == 0  ||
+               strcmp(key, "revoke") == 0       ||
+               strcmp(key, "rotate") == 0       ||
+               strcmp(key, "create") == 0       ||
+               strcmp(key, "ro") == 0) {
+        if (!val || strcmp(val, "true") != 0)
+            krb5_set_error_message(r->context, ret = EINVAL,
+                                   "get-keys \"%s\" q-param accepts "
+                                   "only \"true\"", key);
+        else if (strcmp(key, "materialize") == 0)
+            r->materialize = 1;
+        else if (strcmp(key, "revoke") == 0)
+            r->revoke = 1;
+        else if (strcmp(key, "rotate") == 0)
+            r->rotate = 1;
+        else if (strcmp(key, "create") == 0)
+            r->create = 1;
+        else if (strcmp(key, "ro") == 0)
+            r->ro = 1;
+        if (ret == 0)
+            heim_audit_addkv((heim_svc_req_desc)r, KDC_AUDIT_VIS,
+                             "requested_option", "%s", key);
+    } else if (strcmp(key, "dNSName") == 0 && val) {
+        s = heim_string_create(val);
+        if (!s)
+            ret = krb5_enomem(r->context);
+        else
+            ret = heim_array_append_value(r->hostnames, s);
+        heim_audit_addkv((heim_svc_req_desc)r, KDC_AUDIT_VIS,
+                         "requested_dNSName", "%s", val);
+        ret = hx509_request_add_dns_name(r->context->hx509ctx, r->req, val);
+    } else if (strcmp(key, "service") == 0 && val) {
+        s = heim_string_create(val);
+        if (!s)
+            ret = krb5_enomem(r->context);
+        else
+            ret = heim_array_append_value(r->service_names, s);
+        heim_audit_addkv((heim_svc_req_desc)r, KDC_AUDIT_VIS,
+                         "requested_service", "%s", val);
+    } else if (strcmp(key, "enctypes") == 0 && val) {
+        r->enctypes = strdup(val);
+        if (!(r->enctypes = strdup(val)))
+            ret = krb5_enomem(r->context);
+        heim_audit_addkv((heim_svc_req_desc)r, KDC_AUDIT_VIS,
+                         "requested_enctypes", "%s", val);
+    } else if (strcmp(key, "spn") == 0 && val) {
+        krb5_principal p = NULL;
+        const char *hostname = "";
+
+        heim_audit_addkv((heim_svc_req_desc)r, KDC_AUDIT_VIS,
+                         "requested_spn", "%s", val);
+
+        ret = krb5_parse_name(r->context, val, &p);
+        if (ret == 0) {
+            if (krb5_principal_get_num_comp(r->context, p) < 2 ||
+                krb5_principal_get_num_comp(r->context, p) > 3)
```

Comment:
```suggestion
            if (krb5_principal_get_num_comp(r->context, p) != 2)
```
Until we know how to authorize the domain part of domain-based names, let's not do this.